File: backend/app/schemas/progress.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


# ============================================================================
# User Progress Schemas
# ============================================================================

class UserProgressBase(BaseModel):
    user_id: int
    topic_id: int
    completed_exercises: int = 0
    total_exercises: int = 0
    accuracy_percentage: float = 0.0


class UserProgress(UserProgressBase):
    id: int
    progress_percentage: float
    last_updated: datetime
    
    model_config = {"from_attributes": True}


class UserProgressWithTopic(UserProgress):
    topic_title: str
    topic_description: Optional[str] = None


# ============================================================================
# Lesson Progress Schemas
# ============================================================================

class LessonProgressBase(BaseModel):
    user_id: int
    lesson_id: int
    completed_exercises: int = 0
    total_exercises: int = 0
    completed: bool = False


class LessonProgress(LessonProgressBase):
    id: int
    progress_percentage: float
    last_updated: datetime
    
    model_config = {"from_attributes": True}


class LessonProgressWithDetails(LessonProgress):
    lesson_title: str
    lesson_content: Optional[str] = None


# ============================================================================
# Dashboard Statistics
# ============================================================================

class DashboardStats(BaseModel):
    """Overall student statistics for dashboard"""
    total_exercises_completed: int
    total_exercises_attempted: int
    accuracy_percentage: float
    topics_started: int
    topics_completed: int
    total_topics_available: int
    lessons_started: int
    lessons_completed: int
    total_lessons_available: int
    recent_activity: List[str] = []


class XpSummary(BaseModel):
    """Current XP and level snapshot for the student dashboard."""
    user_id: int
    level: int
    total_xp: int
    current_level_xp: int
    next_level_xp: int
    xp_into_level: int
    xp_to_next_level: int
    progress_percentage: float
    streak_days: int
    streak_multiplier: float
    today_xp: int


# ============================================================================
# Topic Progress Response
# ============================================================================

class TopicProgressSummary(BaseModel):
    """Topic-level progress summary"""
    topic_id: int
    title: str
    description: Optional[str] = None
    grade_number: int
    progress_percentage: float
    accuracy: float
    completed_exercises: int
    total_exercises: int
    lessons_completed: int
    total_lessons: int
    needs_practice: bool = False


# ============================================================================
# Lesson Progress Response
# ============================================================================

class LessonProgressSummary(BaseModel):
    """Lesson-level progress summary"""
    lesson_id: int
    title: str
    progress_percentage: float
    completed_exercises: int
    total_exercises: int
    completed: bool


# ============================================================================
# Recommendations
# ============================================================================

class WeakTopic(BaseModel):
    """Topic that needs practice"""
    topic_id: int
    title: str
    accuracy: float
    reason: str


class RecommendedLesson(BaseModel):
    """Recommended lesson to practice"""
    lesson_id: int
    topic_id: int
    lesson_title: str
    topic_title: str
    reason: str


class ProgressRecommendations(BaseModel):
    """Recommendations for student improvement"""
    weak_topics: List[WeakTopic]
    recommended_lessons: List[RecommendedLesson]
    encouragement_message: str


# ============================================================================
# Freemium / Subscription
# ============================================================================

class UserLimitInfo(BaseModel):
    """Current user usage and limits for freemium system"""
    plan: str  # "free" or "premium"
    
    # AI Exercises
    ai_exercises_remaining: int
    ai_exercises_limit: int
    ai_exercises_used_today: int
    
    # AI Chat
    ai_chat_remaining: int
    ai_chat_limit: int
    ai_chat_used_today: int
    
    # NVO Exams
    nvo_exams_remaining: int
    nvo_exams_limit: int
    nvo_exams_used_today: int
    
    # Image Scans
    image_scans_remaining: int
    image_scans_limit: int
    image_scans_used_today: int
    
    # Premium info
    is_premium: bool
    can_upgrade: bool
    days_until_reset: int

File: frontend/src/services/progress.ts
import apiClient from './api';

// ============================================================================
// Progress API Functions
// ============================================================================

export interface DashboardStats {
  total_exercises_completed: number;
  total_exercises_attempted: number;
  accuracy_percentage: number;
  topics_started: number;
  topics_completed: number;
  total_topics_available: number;
  lessons_started: number;
  lessons_completed: number;
  total_lessons_available: number;
  recent_activity: string[];
}

export interface XpSummary {
  user_id: number;
  level: number;
  total_xp: number;
  current_level_xp: number;
  next_level_xp: number;
  xp_into_level: number;
  xp_to_next_level: number;
  progress_percentage: number;
  streak_days: number;
  streak_multiplier: number;
  today_xp: number;
}

export interface TopicProgress {
  topic_id: number;
  title: string;
  description?: string;
  grade_number: number;
  progress_percentage: number;
  accuracy: number;
  completed_exercises: number;
  total_exercises: number;
  lessons_completed: number;
  total_lessons: number;
  needs_practice: boolean;
}

export interface LessonProgress {
  lesson_id: number;
  title: string;
  progress_percentage: number;
  completed_exercises: number;
  total_exercises: number;
  completed: boolean;
}

export interface WeakTopic {
  topic_id: number;
  title: string;
  accuracy: number;
  reason: string;
}

export interface RecommendedLesson {
  lesson_id: number;
  topic_id: number;
  lesson_title: string;
  topic_title: string;
  reason: string;
}

export interface ProgressRecommendations {
  weak_topics: WeakTopic[];
  recommended_lessons: RecommendedLesson[];
  encouragement_message: string;
}

/**
 * Get overall dashboard statistics
 */
export const getDashboardStats = async (): Promise<DashboardStats> => {
  const response = await apiClient.get('/progress/dashboard');
  return response.data;
};

/**
 * Get XP and level summary for the current user
 */
export const getXpSummary = async (): Promise<XpSummary> => {
  const response = await apiClient.get('/progress/xp-summary');
  return response.data;
};

/**
 * Record a daily activity/login event — updates streak and returns updated XP summary
 */
export const recordActivity = async (): Promise<XpSummary> => {
  const response = await apiClient.post('/progress/record-activity');
  return response.data;
};

// ============================================================================
// Badge API
// ============================================================================

export interface UserBadge {
  key: string;
  title: string;
  emoji: string;
  description: string;
  unlocked_at: string;
}

export const getUserBadges = async (): Promise<UserBadge[]> => {
  const response = await apiClient.get('/progress/badges');
  return response.data;
};

export interface DailyMission {
  id: string;
  title: string;
  description: string;
  duration: string;
  difficulty: string;
  xp_base: number;
  xp_bonus: number;
  emoji: string;
  route: string;
  mission_type: string;
  topic_id?: number;
  lesson_id?: number;
  target_count?: number;
  completed_count?: number;
  is_completed?: boolean;
}

export const getDailyMissions = async (): Promise<DailyMission[]> => {
  const response = await apiClient.get('/progress/daily-missions');
  return response.data;
};

export interface MissionTrackResult {
  mission_id: string;
  target_count: number;
  completed_count: number;
  is_completed: boolean;
  xp_earned: number;
  ignored?: boolean;
  reason?: string;
}

export const trackMissionProgress = async (
  mission_id: string,
  exercise_id: number,
  is_correct: boolean,
): Promise<MissionTrackResult> => {
  const response = await apiClient.post('/progress/daily-missions/track', null, {
    params: { mission_id, exercise_id, is_correct },
  });
  return response.data;
};

// ============================================================================
// Activity Feed
// ============================================================================

export interface ActivityEvent {
  id: number;
  source_type: string;
  xp_amount: number;
  reason: string;
  created_at: string;
}

export const getActivityFeed = async (limit = 20): Promise<ActivityEvent[]> => {
  const response = await apiClient.get('/progress/activity-feed', {
    params: { limit },
  });
  return response.data;
};

/**
 * Get progress for all topics
 */
export const getTopicProgress = async (): Promise<TopicProgress[]> => {
  const response = await apiClient.get('/progress/topics');
  return response.data;
};

/**
 * Get progress for lessons in a specific topic
 */
export const getLessonProgress = async (topicId: number): Promise<LessonProgress[]> => {
  const response = await apiClient.get(`/progress/lessons/${topicId}`);
  return response.data;
};

/**
 * Get personalized recommendations
 */
export const getRecommendations = async (): Promise<ProgressRecommendations> => {
  const response = await apiClient.get('/progress/recommendations');
  return response.data;
};
